src/time_utils.c

```diff
@@ -468,15 +460,25 @@ ts_time_get_noend_or_max(Oid timetype)
 int64
 ts_time_saturating_add(int64 timeval, int64 interval, Oid timetype)
 {
-	if (TS_TIME_IS_INTEGER_TIME(timetype))
-	{
-		int64 time_max = ts_time_get_max(timetype);
+	int64 min_overflow_result;
+	int64 max_overflow_result;
 
-		if (timeval > (time_max - interval))
-			return time_max;
+	if (IS_TIMESTAMP_TYPE(timetype))
+	{
+		min_overflow_result = ts_time_get_nobegin(timetype);
+		max_overflow_result = ts_time_get_noend(timetype);
 	}
-	else if (timeval >= (ts_time_get_end(timetype) - interval))
-		return ts_time_get_noend(timetype);
+	else
+	{
+		min_overflow_result = ts_time_get_min(timetype);
+		max_overflow_result = ts_time_get_max(timetype);
+	}
+
+	if (timeval > 0 && interval > 0 && timeval > (ts_time_get_max(timetype) - interval))
+		return max_overflow_result;
+
+	if (timeval < 0 && interval < 0 && timeval < (ts_time_get_min(timetype) - interval))
+		return min_overflow_result;
```

Comment:
You actually only need to compute the max and min values for these cases, so you can remove the code above and just use this:
```suggestion
	if (timeval > 0 && interval > 0 && timeval > (ts_time_get_max(timetype) - interval))
		return ts_time_get_noend_or_max(timetype);

	if (timeval < 0 && interval < 0 && timeval < (ts_time_get_min(timetype) - interval))
		return ts_time_get_nobegin_or_min(timetype);
```

Reply:
The latter function didn't exist, but I guess I can create it.